[Navigation] Schedule navigate event until after back/forward/traverseTo

rwlbuis opened the pull request:
#### 4735fd370a76616fd6d29dfdd5e1144a928cdb10
<pre>
[Navigation] Schedule navigate event until after back/forward/traverseTo
<a href="https://bugs.webkit.org/show_bug.cgi?id=276546">https://bugs.webkit.org/show_bug.cgi?id=276546</a>

Reviewed by Alex Christensen.

The traverse is supposed to use the session history traversal queue [1]. WebKit does not seem to have this data structure implemented, but
seeing history.back() etc. should work similarly as navigation.back(), we can emulate that and use NavigationScheduler, which puts the navigation
on a timer.

[1] <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal">https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal</a> (Step 12)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-back-forward-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-getState-back-forward-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/back-forward-multiple-frames-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/forward-to-pruned-entry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-detach-same-document-before-navigate-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-repeated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/currententrychange-before-popstate-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/same-document-away-and-back-navigation-api-expected.txt:
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledHistoryNavigationByKey::ScheduledHistoryNavigationByKey):
(WebCore::ScheduledHistoryNavigationByKey::~ScheduledHistoryNavigationByKey):
(WebCore::NavigationScheduler::scheduleHistoryNavigationByKey):
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::abortOngoingNavigation):
* Source/WebCore/page/Navigation.h:

Canonical link: <a href="https://commits.webkit.org/282699@main">https://commits.webkit.org/282699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3782d8de987c13955c911449b99327190b1da55f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63990 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43347 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68012 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14598 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66110 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14878 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51534 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10081 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67059 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40129 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55383 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32153 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12769 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58759 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13098 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69711 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7937 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12623 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58854 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7970 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55482 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59001 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14150 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6587 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39167 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40246 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41429 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39989 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebCore/loader/NavigationScheduler.cpp:
```diff
@@ -315,6 +316,60 @@ class ScheduledHistoryNavigation : public ScheduledNavigation {
+class ScheduledHistoryNavigationByKey : public ScheduledNavigation {
+public:
+    explicit ScheduledHistoryNavigationByKey(const String& key, CompletionHandler<void(ScheduleHistoryNavigationResult)>&& completionHandler)
+        : ScheduledNavigation(0, LockHistory::No, LockBackForwardList::No, false, true)
```
Currently this hardcodes `wasDuringLoad` to `false`.It should be possible to copy the logic from the other scheduled classes to set this accurately.

EDIT: This only seems to be used for redirects, so not relevant to this change. Maybe it is worth the value just being correct anyway.

rwlbuis replied:
I see your point, but according to the spec this should work a lot like history.back, so I prefer emulating that as much as possible.

TingPing reviewed (commented):
Looks excellent.

achristensen07 commented on Source/WebCore/loader/NavigationScheduler.cpp:
```diff
@@ -315,6 +316,60 @@ class ScheduledHistoryNavigation : public ScheduledNavigation {
     Ref<HistoryItem> m_historyItem;
 };
 
+class ScheduledHistoryNavigationByKey : public ScheduledNavigation {
```
I think this should have a comment with a link to https://html.spec.whatwg.org/multipage/browsing-the-web.html#she-navigation-api-key so we know what key it's talking about.

rwlbuis replied:
Makes sense, added.

achristensen07 commented on Source/WebCore/loader/NavigationScheduler.cpp:
```diff
@@ -315,6 +316,60 @@ class ScheduledHistoryNavigation : public ScheduledNavigation {
+        }
+
+        // go(i!=0) from a frame navigates into the history of the frame only,
+        // in both IE and NS (but not in Mozilla). We can't easily do that.
```
Netscape didn't implement the new navigation API.  I think we should remove this comment.  Same for the comment on 354.  If it is intended to share code, it should be shared rather than copied.

rwlbuis replied:
I removed the comments.

rwlbuis committed:
[Navigation] Schedule navigate event until after back/forward/traverseTo
https://bugs.webkit.org/show_bug.cgi?id=276546

Reviewed by Alex Christensen.

The traverse is supposed to use the session history traversal queue [1]. WebKit does not seem to have this data structure implemented, but
seeing history.back() etc. should work similarly as navigation.back(), we can emulate that and use NavigationScheduler, which puts the navigation
on a timer.

[1] https://html.spec.whatwg.org/multipage/nav-history-apis.html#performing-a-navigation-api-traversal (Step 12)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-back-forward-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-getState-back-forward-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/back-forward-multiple-frames-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/forward-to-pruned-entry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-detach-same-document-before-navigate-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-repeated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/currententrychange-before-popstate-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/same-document-away-and-back-navigation-api-expected.txt:
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledHistoryNavigationByKey::ScheduledHistoryNavigationByKey):
(WebCore::ScheduledHistoryNavigationByKey::~ScheduledHistoryNavigationByKey):
(WebCore::NavigationScheduler::scheduleHistoryNavigationByKey):
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
(WebCore::Navigation::abortOngoingNavigation):
* Source/WebCore/page/Navigation.h:

Canonical link: https://commits.webkit.org/282699@main
